### PR TITLE
correctly transfer username validation to auth0

### LINKED ISSUE #238: Unable to specify username requirements
### Description

Im unable to configure my Auth0 Connection with username restrictions and even with the `requires_username` flag.

### Affected Resource(s)

* auth0_connection

### Terraform Configuration Files

1. Issue (require a username at all)
```hcl
resource "auth0_connection" "conn" {
  strategy = "auth0"
  ...

  options {
    requires_username = true
  }
}
```

2. Issue (specify username validation)
```hcl
resource "auth0_connection" "conn" {
  strategy = "auth0"
  ...

  options {
    requires_username = true

    validation = {
       username = {
          min  = 3
          max = 20
       }
    }
  }
}
```

Terraform error:
> Inappropriate value for attribute "validation": element "username": string
required.

Using for example `jsonencode` for validation.username ends up in an Auth0 Api Error:
> Error: 400 Bad Request: Payload validation error: 'Expected type object,null but found type string' on property options.validation.username.

### Expected Behavior

- `requires_username` is correctly send to Auth0 API
- Username validation can be specified
- `validation` has the correct type description in [Terraform docs](https://www.terraform.io/docs/providers/auth0/r/connection.html) (validation says its a string)
  - Using string for validation ends up in a terraform error "Inappropriate value for attribute "validation": map of string required."

### Actual Behavior

In my connection "Requires username" is disabled. According to the [Auth0 API](https://auth0.com/docs/connections/references/options-mgmt-api) the `validation` flag is used to specify that a username is required. 

Terraform Provider and Auth0 API have a type mismatch in the `validation` option flag. I'm therefore unable to specify username requirements.

**Note:** I'll try to make some time free to fix this myself. If anyone else is quicker or has any concerns please let me know.

### Steps to Reproduce

1. Try configure username requirement and validation

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by etwillbefine**
### Proposed Changes

* Correctly sending username validation to Auth0

Fixes #238

#### Acceptance Test Output

```
make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.01s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.85s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.16s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.80s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (19.63s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (0.90s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (6.55s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (0.90s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.83s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.67s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.73s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.84s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.97s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.83s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (7.75s)
PASS
coverage: 27.3% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	74.726s	coverage: 27.3% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.262s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.288s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]
...
```

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

#### Note: Contains breaking changes:
- Moved `options.requires_username` to `options.validation` 
- Changed Data Structure of `options.validation`

### CODE REVIEW

**Comment by etwillbefine** [comment]
As suggested in the README `make test` runs successful. But for the CI run (`make testacc ..`)  it fails locally and on Travis, missing Provider configuration. Help appreciated.

**Comment by alexkappa** [comment]
Hi, thanks for giving this a go!

Regarding the tests failing on CI, its kind of expected as PRs wont have access to secrets. You should run them locally though and paste the output in the right section in the description.

As this is a breaking change, we should add a migration so as to not break existing users.

I wasn't able to see any changes to flattening/expanding. Do the right values get passed along to the sdk? I’m on mobile so I might have missed something.

Cheers,
Alex

**Comment by etwillbefine** [comment]
The Auth0 Client supports [`validation`](https://github.com/go-auth0/auth0/blob/master/management/connection.go#L157) already with a generic interface type. 

During Testing I realised that the `requires_username` actually works. Not sure how one of my connection had it set to false and true in Terraform but when testing with a fresh Connection it works. When sending requires_username within the validation object Auth0 returns a Bad Request that this property is not allowed. That means Auth0's Documentation is not correct about requires_username property. https://auth0.com/docs/connections/references/options-mgmt-api

I created the corresponding mapping for validation, removed changes related to `requires_username` and added Acceptance Test Output to the original Description.

File: auth0/resource_auth0_connection.go [hunk 1]
```diff
@@ -181,9 +204,8 @@ var connectionSchema = map[string]*schema.Schema{
 					Description: "Indicates whether or not to allow user sign-ups to your application",
 				},
 				"requires_username": {
-					Type:        schema.TypeBool,
-					Optional:    true,
-					Description: "Indicates whether or not the user is required to provide a username in addition to an email address",
+					Type:     schema.TypeBool,
+					Optional: true,
```
[the remark this answers is not in the document]
Comment (etwillbefine) [reply auth0/resource_auth0_connection.go:208]: added description again

File: auth0/resource_auth0_connection_test.go [hunk 2]
```diff
@@ -110,11 +112,17 @@ resource "auth0_connection" "my_connection" {
...
+				min = 10
+				max = 40
+			}
+		}
+		requires_username = true
 		enabled_database_customization = false
 		brute_force_protection = true
 		import_mode = false
 		disable_signup = false
-		requires_username = true
```
[the remark this answers is not in the document]
Comment (etwillbefine) [reply auth0/resource_auth0_connection_test.go:117]: cant really revert as in same commit. moved the property to its previous position

File: auth0/resource_auth0_connection.go [hunk 3]
```diff
@@ -81,8 +81,31 @@ var connectionSchema = map[string]*schema.Schema{
 		Elem: &schema.Resource{
 			Schema: map[string]*schema.Schema{
 				"validation": {
-					Type:     schema.TypeMap,
-					Elem:     &schema.Schema{Type: schema.TypeString},
+					Type:     schema.TypeList,
```
[the remark this answers is not in the document]
Comment (etwillbefine) [reply auth0/resource_auth0_connection.go:84]: Thats going to take same time as Im new to Go and Terraform Provider Migrations. Will try to provide it within the next few days.

[the remark this answers is not in the document]
Comment (alexkappa) [reply auth0/resource_auth0_connection.go:84]: I understand :) I’m here if you need any help.

For inspiration, you can see the migration of connections schema a while back.

**Comment by etwillbefine** [comment]
I added the migration to move `validation.min|max` to `validation.username.min|max`. I'm not sure if I need to make sure to convert the previous string value to int. `structure_auth0_validation.go` contains already converting to Int value. 

Here is an updated Test Output:
```
make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.35s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.89s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.19s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.97s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (13.65s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (7.05s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (1.66s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (7.01s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.93s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.74s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.64s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.90s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.97s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.73s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (1.69s)
PASS
coverage: 26.9% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	70.714s	coverage: 26.9% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.246s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.146s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]
```

**Comment by etwillbefine** [comment]
Here is an updated Test Output:

make testacc TESTS=TestAccConnection
==> Checking that code complies with gofmt requirements...
?   	github.com/terraform-providers/terraform-provider-auth0	[no test files]
=== RUN   TestAccConnection
--- PASS: TestAccConnection (2.20s)
=== RUN   TestAccConnectionAD
--- PASS: TestAccConnectionAD (0.89s)
=== RUN   TestAccConnectionAzureAD
--- PASS: TestAccConnectionAzureAD (7.25s)
=== RUN   TestAccConnectionOIDC
--- PASS: TestAccConnectionOIDC (1.68s)
=== RUN   TestAccConnectionWithEnbledClients
--- PASS: TestAccConnectionWithEnbledClients (13.82s)
=== RUN   TestAccConnectionSMS
--- PASS: TestAccConnectionSMS (7.09s)
=== RUN   TestAccConnectionEmail
--- PASS: TestAccConnectionEmail (1.63s)
=== RUN   TestAccConnectionSalesforce
--- PASS: TestAccConnectionSalesforce (7.10s)
=== RUN   TestAccConnectionGoogleOAuth2
--- PASS: TestAccConnectionGoogleOAuth2 (0.95s)
=== RUN   TestAccConnectionFacebook
--- PASS: TestAccConnectionFacebook (7.76s)
=== RUN   TestAccConnectionApple
--- PASS: TestAccConnectionApple (1.64s)
=== RUN   TestAccConnectionLinkedin
--- PASS: TestAccConnectionLinkedin (7.75s)
=== RUN   TestAccConnectionGitHub
--- PASS: TestAccConnectionGitHub (0.91s)
=== RUN   TestAccConnectionConfiguration
--- PASS: TestAccConnectionConfiguration (7.66s)
=== RUN   TestAccConnectionSAML
--- PASS: TestAccConnectionSAML (1.68s)
PASS
coverage: 26.9% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0	70.348s	coverage: 26.9% of statements
?   	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/debug	[no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/random	0.254s	coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok  	github.com/terraform-providers/terraform-provider-auth0/auth0/internal/validation	0.081s	coverage: 0.0% of statements [no tests to run]
?   	github.com/terraform-providers/terraform-provider-auth0/version	[no test files]

**Comment by fbarbare** [comment]
Any updates on that PR? We need to configure those options too
@alexkappa

**Comment by alexkappa** [comment]
Thank you @etwillbefine, I've merged your changes in. Will cut a new release soon with the validation fix in place, cheers!

**Pull request merged**